### PR TITLE
Migration to Rocky8 spack-stack installations on Jet

### LINKED ISSUE #2377: Migrate to Rocky8 spack-stack installations on Jet
### What new functionality do you need?

Jet is upgrading to the Rocky8 Linux OS, which requires a new spack-stack installation on the platform.  The global workflow, subcomponents, and external dependencies will need to be recompiled and tested on the platform before full transition in mid-April.

### What are the requirements for the new functionality?

- [x] Spack-stack v1.6.0 installation
   - [x] `/lfs4/HFIP/hfv3gfs/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8`
   - [x] `/lfs4/HFIP/hfv3gfs/role.epic/spack-stack/spack-stack-1.5.1/envs/unified-env-rocky8`
- All submodules transitioned to Rocky8
   - [ ] UFS
   - [ ] GSI
   - [ ] GDAS
   - [ ] GSI-Utils
   - [ ] GSI-Monitor
   - [ ] EMC_verif-global
   - [ ] UFS_Utils
   - [ ] GFS-Utils
- External dependencies need to be rebuilt
   - [ ] Obsproc
   - [ ] Fit2Obs
   - [ ] TC_Tracker

### Acceptance Criteria

The global workflow is able to run in both cycled and forecast-only modes at resolutions up to C384 with all subcomponents and external dependencies running successfully.

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by InnocentSouopgui-NOAA**
# Description

This migrates Global Workflow to Rocky8 spack-stack installations on Jet.
Jet has moved from CentOS7 to Rocky8.

Resolves #2377
Refs NOAA-EMC/UPP#919
Refs NOAA-EMC/gfs-utils#60
Refs NOAA-EMC/GSI#732
Refs NOAA-EMC/GSI-Monitor#130
Refs NOAA-EMC/GSI-utils#33

# Type of change

- Bug fix (fixes something broken)
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
Cycled experiments (48+ hours) at resolutions
- [x] 96/48 on xjet and kjet
- [x] 192/96 on kjet
- [x] 384/192 on kjet

Forecast only experiment (48+ hours) at resolutions
- [x] 48
- [x] 96
- [x] 192
- [x] 384

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by InnocentSouopgui-NOAA** [comment]
Successful cycling over 3 days at resolution 96/48 on xjet and kjet.
Higher resolution in progress.

**Comment by InnocentSouopgui-NOAA** [comment]
@DavidHuber-NOAA
There is a need to update the config files on Jet. I need some guidances here to proceed.
There are two reasons for the need to update the config files:
1. The required memory per node for some jobs is beyond the available memory per node, so those jobs never start; for example:
    - esfc jobs requires 80GB of memory per node, xjet available memory per node is 61GB
    - upp jobs requires 96GB of memory per node, kjet available memory per node is 93GB
When the memory requirement is lowered, those jobs run.

2. The wall clock limit is too short for some jobs at high resolution; for example:
    - enkfgdasesfc: 8 min too short at 192/60 (bump to 15?)
    - gdasfcst: 15 min too short at 192/60 (bump to 20?)
    - sfcanl: 10 min too short at 384/192 (bump to 15?)

The required memory per node is the critical one. Should I add specific instruction like
```
if [[ ${machine} == "JET" && ${partition} == "kjet" ]]; then export memory_upp="80GB" ; fi
if [[ ${machine} == "JET" && ${partition} == "xjet" ]]; then export memory_upp="60GB" ; fi
```
in the config.resources?

For the wall clock, 
```
if [[ ${machine} == "JET" ]]; then export wtime_sfcanl="00:15:00" ; fi # for sfcanl jobs
```
What do you think?

**Comment by DavidHuber-NOAA** [comment]
>     2. The wall clock limit is too short for some jobs at high resolution; for example:
>        
>        * enkfgdasesfc: 8 min too short at 192/60 (bump to 15?)
>        * gdasfcst: 15 min too short at 192/60 (bump to 20?)
>        * sfcanl: 10 min too short at 384/192 (bump to 15?)
These all sound fine to me.  I think you can make these global (not machine-specific), since they are relatively small bumps.  Also, we don't typically test at C192, so these updates may be needed across the board.

> The required memory per node is the critical one. Should I add specific instruction like
> 
> ```
> if [[ ${machine} == "JET" && ${partition} == "kjet" ]]; then export memory_upp="80GB" ; fi
> if [[ ${machine} == "JET" && ${partition} == "xjet" ]]; then export memory_upp="60GB" ; fi
> ```
How about just telling srun to use all available memory for Jet nodes:
```
if [[ ${machine} == "JET" ]]; then export memory_upp="0" ; fi
```

**Comment by InnocentSouopgui-NOAA** [comment]
That sounds good, I will do that.

**Comment by InnocentSouopgui-NOAA** [comment]
gfsfcst task is hanging in forecast-only experiment. I have the end of the log file below in case it rings a bell.
That is about 3 minutes of wall-clock time into the run and it does not not move past that point.
I am running a test at the resolution 48 on Jet.

@aerorahul do you have any idea?

==== section of the log file ===
```
33:  EXPTOTAL is selected
33:  But PDLIB is not
36:  EXPTOTAL is selected
36:  But PDLIB is not
38:  EXPTOTAL is selected
38:  But PDLIB is not
39:  EXPTOTAL is selected
39:  But PDLIB is not
24:             Fields   : no fields defined
24:  
24:   Time interval :  
24:  --------------------------------------------------
24:        Starting time : 2021/03/23 12:00:00 UTC
24:        Ending time   : 2021/03/28 12:00:00 UTC
24:  
24:  
24:   Output requests :  
24:  --------------------------------------------------
24:        No dedicated output process, any file system.
24:  
24:        Type 1 : Fields of mean wave parameters
24:       -----------------------------------------
24:             From     : 2021/03/23 12:00:00 UTC
24:             To       : 2021/03/28 12:00:00 UTC
24:             Interval :            01:00:00
24:  
24:  
24:        Type 2 : Point output                   
24:       -----------------------------------------
24:             From     : 2021/03/23 12:00:00 UTC
24:             To       : 2021/03/28 12:00:00 UTC
24:             Interval :            01:00:00
24:  
24:  
24:        Type 8 : Restart files second request   
24:       -----------------------------------------
24:             From     : 2021/03/23 12:00:00 UTC
24:             To       : 2021/03/28 12:00:00 UTC
24:             Interval :            12:00:0
24:  
24:        Wave model ...
24:  EXPTOTAL is selected
24:  But PDLIB is not
```

**Comment by JessicaMeixner-NOAA** [comment]
@InnocentSouopgui-NOAA this is a  compile time vs runtime mismatch issue in the wave model.  What test are you running when you get this error?

**Comment by InnocentSouopgui-NOAA** [comment]
> @InnocentSouopgui-NOAA this is a compile time vs runtime mismatch issue in the wave model. What test are you running when you get this error?
I am running a forecast-only S2SW experiment using the prototype_ICs.

**Comment by JessicaMeixner-NOAA** [comment]
Prototype ICs, likely mean you are trying to run an unstructured grid for the waves, which means you need to add -w when compiling the model.

**Comment by aerorahul** [comment]
@InnocentSouopgui-NOAA 
When building and running for `system = gfs`  cases, do not use the `-w` option with `build_all.sh`.
The `-w` option should only be used with the `system = gefs` cases where a structured WW3 grid is used.

**Comment by InnocentSouopgui-NOAA** [comment]
@JessicaMeixner-NOAA I recompiled gfs with -w and it is running now without hanging.
However with the comment from @aerorahul about not compiling gfs with -w, Ihave more questions:
The goal here is to test forecast only on Jet/Rocky8 before merging everything and have it available for other people who run on Jet.

What is the best course of action?

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA The meaning of the `-w` flag for `build_all.sh` changed recently in the develop branch.  It used to mean 'use the unstructured wave grid', but now it means 'use the structured wave grid'.  If you sync your branch with develop, you will capture this update.  Once doing so, you _**should not**_ use `-w` when running GFS experiments.

**Comment by InnocentSouopgui-NOAA** [comment]
> @InnocentSouopgui-NOAA The meaning of the `-w` flag for `build_all.sh` changed recently in the develop branch. It used to mean 'use the unstructured wave grid', but now it means 'use the structured wave grid'. If you sync your branch with develop, you will capture this update. Once doing so, you _**should not**_ use `-w` when running GFS experiments.

Great, thanks. I remember seeing a notification about the change in the meaning of -w flag.

**Comment by InnocentSouopgui-NOAA** [comment]
@DavidHuber-NOAA, Pretty much everything is ready to submit this pull request. I am waiting for the last forecast only experiment to complete.
Question: who maintains the external dependencies on Jet?
TC_tracker, OBSPROC and Fit2Obs were compiled with with stack spack 1.6.0 and link from my personal directory for all tests.
Fit2Obs source code was Downloaded from https://github.com/NOAA-EMC/Fit2Obs/archive/refs/tags/wflow.1.1.0.tar.gz. TC_tracker and OBSPROC source code was copied from the official BASE_GIT on Jet and recompiled.

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA Those are maintained by @KateFriedman-NOAA on Jet.  Kate, would you be able to get Innocent's branches installed into the official locations?

Innocent, when that's finished, you will want to update the locations in your branch.  I believe they will end up in /lfs4/HFIP/hfv3gfs/glopara/git.

**Comment by KateFriedman-NOAA** [comment]
@InnocentSouopgui-NOAA Can you point me to your branches for each of these repos? It's a bit of a mixed bag for each external piece:

Fit2Obs:
- latest install on Jet: /lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.1.0
- Tag on GitHub: https://github.com/NOAA-EMC/Fit2Obs/releases/tag/wflow.1.1.0
- We need to roll any changes you made into the Fit2Obs `develop` branch and make a new tag (which I would then install everywhere)
- An issue needs to be opened in Fit2Obs as well (to document the Jet update): https://github.com/NOAA-EMC/Fit2Obs/issues

OBSPROC (actually it's the prepobs repo, we don't use any execs from the obsproc repo so I only update that when there is a science/script change, OBSPROC is the umbrella name for the obsproc, prepobs, other related repos that we don't need; we just use obsproc and prepobs):
- need to get Jet Rocky8 updates into our branch with prepobs via PR: https://github.com/NOAA-EMC/prepobs/tree/dev/gfsv17
- then I can install the `dev/gfsv17` branch on Jet (we haven't installed it there yet since the prior install still worked before)

TC_tracker:
- similar to prepobs, we are operating out of a branch currently: https://github.com/NOAA-EMC/TC_tracker/tree/feature/GFSv17_com_reorg
- need to get updates into that branch via PR

Once the above tags/branches are updated I can install them in the glopara space on Jet. Need to test with personal branches until then. Let me know how I can assist with getting updates into those repos.

**Comment by KateFriedman-NOAA** [comment]
I should also say...you can make me the reviewer and/or assignee on any PRs for those three repos. I can review, merge, and tag in those repos.

**Comment by InnocentSouopgui-NOAA** [comment]
@KateFriedman-NOAA
I will open the issue for those external dependencies and bring the changes in the repos. Since repositories were so outdated, still using stack-spack, I copied the versions already in /lfs4/HFIP/hfv3gfs/glopara/git and recompiled with spack-stack. And downloaded the tarball of Fit2Obs that matched what was in /lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.1.0.

**Comment by aerorahul** [comment]
> @KateFriedman-NOAA I will open the issue for those external dependencies and bring the changes in the repos. Since repositories were so outdated, still using stack-spack, I copied the versions already in /lfs4/HFIP/hfv3gfs/glopara/git and recompiled with spack-stack. And downloaded the tarball of Fit2Obs that matched what was in /lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.1.0.

@InnocentSouopgui-NOAA 
It is on our TODO to make obsproc, prepobs, fit2obs, and tracker repos as submodules.

**Comment by KateFriedman-NOAA** [comment]
> I will open the issue for those external dependencies and bring the changes in the repos.

Sounds good, thanks @InnocentSouopgui-NOAA ! Let me know if you need me at all ahead of opening PRs.

**Comment by InnocentSouopgui-NOAA** [comment]
@KateFriedman-NOAA for prepobs, it looks like there other steps in addition to build to get everything in  place.

Let's break it down a little more.

I cloned the prepobs repository (dev/gfsv17 branch), built and used it to replace the opbproc, and the prep tasks are failing because of missing scripts. I looked through the logs and see that there are files missing from the repository.

Before the migration to rocky8, /mnt/lfs4/HFIP/hfv3gfs/glopara/git/obsproc/v1.1.2 is used in global workflow.
If I understood your explanation above, that obsproc is the prepobs repo then, I am missing few steps to get everything in place. For instance, those sub-directories are missing:
 - jobs: a set of scripts used by some jobs
 - parm: set of parameters files or so.
 - exec : set of executables of so.
 - docs
 - ecf
 - triggers

Thoughts?

**Comment by KateFriedman-NOAA** [comment]
@InnocentSouopgui-NOAA Apologies for any confusion. Let me be more clear about obsproc/prepobs:

We use both obsproc and prepobs repos (under the uppercase OBSPROC umbrella of repos) but we only use execs from prepobs so we've only had to update prepobs for ports/libraries. There will be an obsproc update in ops soon (to v1.2) but I'll handle that install when it is ready.

You can leave the obsproc setting in `config.base` as is. You'll need to change the prepobs module path in `modulefiles/module_base.jet.lua`. Here: https://github.com/NOAA-EMC/global-workflow/blob/develop/modulefiles/module_base.jet.lua#L45

Point to your test copy there and then when I install the official copy under the glopara space we'll need to change the `feature-GFSv17_com_reorg_log_update` to the new install name. It won't have a version # yet until our changes make it into the version that goes into ops (down the road) so we'll have a non-versioned name for now.

Let me know if you have any questions. :)

**Comment by InnocentSouopgui-NOAA** [comment]
@KateFriedman-NOAA ,

Thank you for the clarifications. let's give it another iteration.
Now that you pointed me to `modulefiles/module_base.jet.lua` I did not account for that one. It is working just fine as it is. Possibly the executables were linked statically. Since it is working just fine, we can keep it as is for the time being.

For now, only the tasks taking the settings from config.* are using the recompiled external modules, namely:
 - obsproc setting in config.base
 - TC_tracker setting in config.tropcy

The the following two are working just fine without being recompiled:
- Fit2Obs (/lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.0.0) the version used is 1.0.0
- prepobs (/lfs4/HFIP/hfv3gfs/glopara/git/prepobs/feature-GFSv17_com_reorg_log_update)

I am tempted to test with the existing TC_tracker and obsproc in /lfs4/HFIP/hfv3gfs/glopara/git ; If it works just fine as Fit2Obs and prepobs, we can keep all those external dependencies as is for now, and migrate them later. I can keep the draft pull requests in place.

Your thoughts!

**Comment by KateFriedman-NOAA** [comment]
> I am tempted to test with the existing TC_tracker and obsproc in /lfs4/HFIP/hfv3gfs/glopara/git ; If it works just fine as Fit2Obs and prepobs, we can keep all those external dependencies as is for now, and migrate them later.

@InnocentSouopgui-NOAA If the rest of the system (particularly the critical components like GSI, UWM, UPP, UFS_UTILS, etc.) work in your branch and the current installed versions of those externals still work post-Rocky8-transition then I agree...let's finish them up later and get the needed Rocky8 updates in for the critical parts of the system.

Let me know if the current TC_tracker and obsproc work in your testing. Thanks!

**Comment by InnocentSouopgui-NOAA** [comment]
> Let me know if the current TC_tracker and obsproc work in your testing. Thanks!
@KateFriedman-NOAA 
- obsproc works just fine on one cycle
- TC_tracker is broken because of netcdf libraries and a couple of problems in other scripts.

I am doing few more tests and the I will all the pull request, but the only one that will need to be re-installed before the global workflow pull request is TC_tracker

**Comment by InnocentSouopgui-NOAA** [comment]
@KateFriedman-NOAA,
Are you done with the version upgrade of fit2obs everywhere?
That way, I can add it to this pull request.

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA, Are you done with the version upgrade of fit2obs everywhere? That way, I can add it to this pull request.

Yes, please proceed. @WalterKolczynski-NOAA is waiting for Orion to be back and then will install the new Fit2Obs tag there. Don't wait for that now, please proceed.

**Review by DavidHuber-NOAA — COMMENTED** [review]
Do any submodules need to be updated for this?

**Comment by InnocentSouopgui-NOAA** [comment]
> Do any submodules need to be updated for this?

All the submodules have been updated and the pull requests merged.
As of now there is only one external dependency (TC_tracker) that is still waiting for review.

**Comment by aerorahul** [comment]
@InnocentSouopgui-NOAA Can you please resolve the conflicts?

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources]: What happens if it get "some" memory?

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources]: What happens if it get "some" memory?
Would it be better to cast it as:
```
if [[ ${machine} != "JET" ]]; then export memory_esfc="80GB" ; fi
```

Comment (InnocentSouopgui-NOAA) [reply parm/config/gfs/config.resources]: > What happens if it get "some" memory? Would it be better to cast it as:
> 
> ```
> if [[ ${machine} != "JET" ]]; then export memory_esfc="80GB" ; fi
> ```

That will not work on xjet partition; Xjet has a max of 64G/node.

File: parm/config/gfs/config.resources
Comment (InnocentSouopgui-NOAA) [reply parm/config/gfs/config.resources]: As in the case of esfc, xjet and kjet have very different size of memory per nodes. Setting 0 (to use all available memory) made sense to keep the modifications to minimal and allow higher resolution to run on kjet while lower res can run on xjet.

Comment (aerorahul) [reply parm/config/gfs/config.resources]: would `unset memory_upp` work instead of `export memory_upp=0`?

Comment (InnocentSouopgui-NOAA) [reply parm/config/gfs/config.resources]: I don't know if `unset memory_upp` would work or not. I have not used that option before. Let me tag @DavidHuber-NOAA to see if he tried that before.

File: parm/config/gfs/config.resources
Comment (WalterKolczynski-NOAA) [reply parm/config/gfs/config.resources]: That would specifically be if it wasn't on Jet. On Jet the value would be unset.

Comment (InnocentSouopgui-NOAA) [reply parm/config/gfs/config.resources]: My bad, I read incorrectly. I get it now.
I will set a run with that and see how it goes.
Same with memory_upp, I will test both together.

**Comment by InnocentSouopgui-NOAA** [comment]
> @InnocentSouopgui-NOAA Can you please resolve the conflicts?

I just resolved the conflicts. I am now testing unsetting variables on JET instead on setting it to zero.

File: parm/config/gfs/config.resources
Comment (InnocentSouopgui-NOAA) [reply parm/config/gfs/config.resources]: @aerorahul I changed to `unset memory_upp` for jet

File: parm/config/gfs/config.resources
Comment (InnocentSouopgui-NOAA) [reply parm/config/gfs/config.resources]: @aerorahul, I changed to your suggestion, setting `memory_esfc` only if not on JET.

**Review by aerorahul — APPROVED** [review]
looks good.

**Comment by WalterKolczynski-NOAA** [comment]
Don't the submodules in workflow need to be updated to the new hashes that include your changes to those repos?

**Comment by InnocentSouopgui-NOAA** [comment]
> Don't the submodules in workflow need to be updated to the new hashes that include your changes to those repos?

Sure they need to be updated, I am not yet familiar with that part.
What do I need to do?

**Comment by DavidHuber-NOAA** [comment]
For each of the submodules, you need to locally check out the versions that you need, then `git add; git commit` them.  For instance, with the GSI, it looks like you updated the Jet modules at https://github.com/NOAA-EMC/GSI/commit/8e279f9c734097f673b07e80f385b2623d13ba4a, so you would:
```bash
cd <global-workflow>/sorc/gsi.fd
git fetch
git checkout 8e279f9c
cd ../
git add gsi.fd
git commit -m "Update gsi hash to include Jet's Rocky8 upgrade"
```

**Comment by KateFriedman-NOAA** [comment]
@InnocentSouopgui-NOAA In addition to updating the component submodules (thanks @DavidHuber-NOAA for providing those instructions!), the `fit2obs_ver` in `versions/run.spack.ver` needs to be updated to `1.1.0`. See this line to make the change:
https://github.com/NOAA-EMC/global-workflow/blob/develop/versions/run.spack.ver#L35

Are we still waiting on the TC_tracker updates to go in? If so, we can take care of that afterwards, let's get the submodules and such wrapped up and merge this PR. Thanks!

**Comment by InnocentSouopgui-NOAA** [comment]
@KateFriedman-NOAA,
What about  `fit2obs` version `1.1.1` you mentioned last week or so? Is it not ready yet?

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA, What about `fit2obs` version `1.1.1` you mentioned last week or so? Is it not ready yet?

It's ready, sorry I mentioned `1.1.0` and should have said `1.1.1`. The `1.1.1` is installed. You can proceed with updating `fit2obs_ver=1.1.1`.

**Comment by InnocentSouopgui-NOAA** [comment]
Submodules updated, fit2obs version updated as well.

**Review by KateFriedman-NOAA — APPROVED** [review]
Updates for Rocky8 on Jet look good. Thanks for your work on this @InnocentSouopgui-NOAA !

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

**Review by DavidHuber-NOAA — APPROVED** [review]
Changes look good.  Thanks!

**Pull request merged**
